File: python/binary-search/binary_search.py
def find(search_list, value):
    lower_bound, upper_bound = 0, len(search_list) - 1
    while upper_bound >= lower_bound:
        index = (upper_bound + lower_bound) // 2
        if value > search_list[index]:
            lower_bound = index + 1
        elif value < search_list[index]:
            upper_bound = index - 1
        else:
            return index
    raise ValueError("value not in array")
            

File: cpp/binary-search/binary_search.cpp
#include "binary_search.h"
#include <vector>
#include <stdexcept>

namespace binary_search {
    int find(std::vector<int> array, int value) {
        int result = position(array, value);
        if (result == -1) throw std::domain_error("could not find value");
        else return result;
    }

    int position(std::vector<int> array, int value) {
        int size = array.size();
        if (size == 0) return -1;
        if (array[size / 2] == value) return size / 2;
        else if (array[size / 2] > value) {
            return position({array.begin(), array.begin() + (size / 2)}, value);
        } else {
             int index = position({array.begin() + (size / 2)+1, array.end()}, value);
             if (index != -1) return size / 2 + 1 + index;
             else return -1;
        }
    }
}  // namespace binary_search
